### ISSUE TITLE
Dataset creates huge Array when run WideAndDeep Model on 100gb data and 1 million features

### ISSUE THREAD

**Issue by bethunebtj** [intel-analytics/analytics-zoo#695]
This maybe a BigDL issue , but I came across this problem when I tried to run W&D model , thus I decide to submit it here.

Here is the problem :  when I start train model on spark , I always fail on huge data and large features. Spark executor keep on running out of memory even when I set the heap memory / direct memory very large. And I notice that the storage memory used is quite low.  I'm skilled on spark, so please believe me I konw how to deal with OOM on spark.

For instance, I assigned 2048GB memory total to the application , and data cached (rdd names are "training rdd" & "thread models") only occupies 100GB memory. 

Then the executor fails , telling me that **it cannot find a rdd file** like this
 `spark .FetchFailedException: Failure while fetching StreamChunkId no such file exception
` , 
and has to re-calculate the rdd , and it fails again. **The executor log shows that the MemoryStorage tries to store a single partition rdd of about 50GB , while the executor memory cannot cache such a big file in memory**. 

I checked the source code and found two places: 

1. First relate to cache: 
```  
//DataSet.scala 
override def cache(): Unit = {
    buffer.count()
    indexes.count()
    isCached = true
  }
```
when dataset tries to cache data , it simply calls an **action** and it triggers the code below:
```
  def rdd[T: ClassTag](data: RDD[T]): DistributedDataSet[T] = {
    val nodeNumber = Engine.nodeNumber()
    new CachedDistriDataSet[T](
      data.coalesce(nodeNumber, true)
        .mapPartitions(iter => {
          Iterator.single(iter.toArray)
        }).setName("cached dataset")
        .cache()
    )
  }
```
It calls the cache() method of spark ,which will just cache the data in memory ,even if the executor memory cannot hold the partition. **I believe that's the first reason why we came across the loss of the rdd and OOM exception.** To avoid this , we should just replace all `cache()` in dataset to `persist(MEMORY_AND_DISK)` or `persist(MEMORY_AND_DISK_SER)`. This may cause the application to run more slowly ,but **better slow than error** ,isn't it?

2. The second place relates to the deeper problem : why will my application create an object of 50GB ? That's really too big . 
```
  def rdd[T: ClassTag](data: RDD[T]): DistributedDataSet[T] = {
    val nodeNumber = Engine.nodeNumber()
    new CachedDistriDataSet[T](
      data.coalesce(nodeNumber, true)
        .mapPartitions(iter => {
          Iterator.single(iter.toArray)
        }).setName("cached dataset")
        .cache()
    )
  }
```
Same code here. We can see that when I run my W&D application, I have about 100 million train `Sample` , and **the code in `Iterator.single(iter.toArray)` turns the whole rdd into one big Array**. The array is of couse huge , for it contains 100 million objects.

I just can't understand why we convert the distributed rdd into one big array , Just for zip  with immutable index?  We can assign index with distributed rdds , and other actions like shuffle. Is it just a bug ?

Sorry for cannot provide a demo for this issue, besides.

**Comment by hhbyyh** [comment]
Thanks for posting the issue. Can I know the partition number, executor memory and executor cores of your application? Basically I want to figure out how much memory is available for each of the Spark task and how much data is there in a single partition.

**Comment by bethunebtj** [comment]
@hhbyyh 
Thx for concern. I used 100 executors ,with each assigned 1 core , 20GB memory and 4GB off-heap memory . The input data is divided into 100 files , thus I set the executor number to 100. The input data contains about 20 millon lines.

And then I use `spark.read.text().repartition(800)` to try to decrease the partition size ( I thought it may help to split the records to smaller arrays) , but it doesn't work.

When the code runs to this line ,the spark ui tells me the whole rdd to deal with has 100 partitions , while in  previous stages it shows that the data has 800 partitions. Thus the only way to decrease the array length is to allocate more executors.

**Comment by bethunebtj** [comment]
Hi yuhao, I checked the source code and found that if I just modify this line ` val nodeNumber = Engine.nodeNumber()` in `DataSet.scala` or just delete `coalesce(nodeNumber, true)` , my rdd partition settings may take affect, I'll try it out in a few days.

Thus I still want to know why dataset transforms the rdd into one big array : `Iterator.single(iter.toArray)`, it is quite unfriendly to jvm. If its goal is to make matrix calculation  easier , will it be better  if we can adjust the array length and default partition number according to ( **batch size / core size**)? Is there something I didn't get?

**Comment by zhichao-li** [comment]
@bethunebtj  seems like you've already figured out the logic. The reason why we cached the array is only for the performance purpose as Analytic-Zoo need to shuffle and retrieve mini-batch data for training after each epoch (We don't want to involve any IO here and a dedicated async thread and queue for this might also be an option). Analytic-zoo train the model in distributed fashion via `data parallel` , so the sourcing data is re-partitioned to the same size of number executors before going into training stage.

**Comment by helenlly** [comment]
fix in BigDL#2679
